Remove deprecated codepath for old-style autograd.Function (#30696)

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33956 Remove deprecated codepath for old-style autograd.Function (#30696)**

Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

Commit:
Remove deprecated codepath for old-style autograd.Function (#30696)

[ghstack-poisoned]

Commit:
Update on "Remove deprecated codepath for old-style autograd.Function (#30696)"


Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

[ghstack-poisoned]

Commit:
Update on "Remove deprecated codepath for old-style autograd.Function (#30696)"


Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -609,69 +609,6 @@ PyObject* process_outputs(PyObject *op_obj, const std::shared_ptr<PyNode>& cdata
 }
 
 // Legacy codepath
```
I think the comment should be removed as well.

albanD commented on test/test_autograd.py:
```diff
@@ -1616,31 +1597,7 @@ def __del__(self):
-        # it.  I'm mostly making sure we don't segfault here.
-        y.backward()
-        self.assertEqual(x2.grad, x2)
+            CollectOnDelete().forward(torch.randn(1, requires_grad=True)).backward()
```
This will raise on error on construction no?

glaringlee replied:
no, it is not get called.

albanD replied:
What is this test doing if this is not called?

glaringlee replied:
No, I mean, __call__ is not triggered in CollectOnDelete().

albanD replied:
Is there any reason why we allow instantiating old-style Function?
You might want to error out in `FunctionMeta`'s `__init__` in function.py as well if you detect an old style Function.

glaringlee replied:
Nope, no spicial reason. I am just not sure whether we should disallow people creating Function obj. People can call forward using CollectOnDelete.forward() or CollectOnDelete.apply() directly, but there is nothing wrong for people doing CollectOnDelete().forward() as well. Please advise.

albanD commented on torch/autograd/function.py:
```diff
@@ -139,8 +139,11 @@ class Function(with_metaclass(FunctionMeta, _C._FunctionBase, _ContextMethodMixi
         >>>         return grad_output * result
     """
 
-    # only for backward compatibility
```
Maybe update the comment above by removing "Each function object is meant to be used only once (in the forward pass)." as it is not relevant.
And add to the code sample:
```
# Use it by calling the apply method:
output = Exp.apply(input)
```

albanD reviewed (commented):
I am surprised there isn't more tests to fix, let's wait for CI.

Commit:
Update on "Remove deprecated codepath for old-style autograd.Function (#30696)"


Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

[ghstack-poisoned]

albanD reviewed (approved):
After an offline discussion.
More things could be removed but they are postponed to a following PR.

Commit:
Update on "Remove deprecated codepath for old-style autograd.Function (#30696)"


Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

[ghstack-poisoned]

Commit:
Update on "Remove deprecated codepath for old-style autograd.Function (#30696)"


Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

[ghstack-poisoned]

Commit:
Update on "Remove deprecated codepath for old-style autograd.Function (#30696)"


Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

[ghstack-poisoned]

Commit:
Update on "Remove deprecated codepath for old-style autograd.Function (#30696)"


Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

[ghstack-poisoned]
